patchwork/scheduling.py

```diff
@@ -220,40 +217,57 @@ class RootQuestionSession(Session):
     # sessions in a real app.
     def __init__(self, scheduler: Scheduler, question: str) -> None:
         super().__init__(scheduler)
-        resulting_context, self.final_answer_promise = scheduler.ask_root_question(question)
-        if resulting_context is None and not self.is_fulfilled():
-            self.current_context = self._choose_next_context()
+        self.current_context, self.final_answer_promise = \
+            scheduler.ask_root_question(question)
+        self.promise_to_advance = self.final_answer_promise
+        self.final_answer = None
+        self.act()
+
+    def choose_promise(self, root: Address) -> Address:
+        """Return unfulfilled promise from hypertext tree with root ``root``.
+
+        Parameters
+        ----------
+        root
+            Address pointing at hypertext or a promise of it.
+        """
+        if not self.sched.db.is_fulfilled(root):
+            return root
+
+        return next((self.choose_promise(child)
+                     for child in self.sched.db.dereference(root).links()),
+                    None)
+
+    # Note: This method is ugly. It's confusing that the action is optional, and
+    # the union return type increases complexity downstream. However, I couldn't
+    # find a better solution and I have to get on with my work. Feel free to
+    # refactor when you feel inspired.
```

Comment:
I think `act` should at least always take an action. (I can think about it if you want to move on to other work, but it should be addressed before we merge the PR.)

Reply:
I don't mind spending more time on it. It's just that I mulled it over for two hours and felt that I wasn't getting things done. This is not a crucial piece of code, anyway. But let me know if you'd like to see an alternative suggestion.

Reply:
Could be good to see your best alternative proposal (but if you've already thought about it for 2h, you're probably right that there are better uses of time than thinking about it more).

Reply:
b93ccac7 is an alternative.